File: typespec/regions/regions.go
package regions

import "github.com/vetchium/src/typespec/common"

type ListSignupRegionsRequest struct {
	ResidentCountry common.CountryCode    `json:"resident_country"`
	PaginationKey   *common.PaginationKey `json:"pagination_key,omitempty"`
}

func (r *ListSignupRegionsRequest) Normalize() {}
func (r ListSignupRegionsRequest) Validate() []string {
	fields := []string{}
	if !common.IsCountryCode(r.ResidentCountry) {
		fields = append(fields, "resident_country")
	}
	if r.PaginationKey != nil && !common.IsPaginationKey(*r.PaginationKey) {
		fields = append(fields, "pagination_key")
	}
	return fields
}

type SignupRegion struct {
	TenantID       string             `json:"tenant_id"`
	HostingCountry common.CountryCode `json:"hosting_country"`
	HubURL         string             `json:"hub_url"`
	Recommended    bool               `json:"recommended"`
}
type ListSignupRegionsResponse struct {
	CatalogVersion    string                `json:"catalog_version"`
	Regions           []SignupRegion        `json:"regions"`
	NextPaginationKey *common.PaginationKey `json:"next_pagination_key"`
}

File: typespec/regions/regions.ts
import type { CountryCode } from "../common/localization.ts";
import { isCountryCode } from "../common/localization.ts";
import type { PaginationKey } from "../common/pagination.ts";
import { isPaginationKey } from "../common/pagination.ts";
export interface ListSignupRegionsRequest {
  resident_country: CountryCode;
  pagination_key?: PaginationKey;
}
export function validateListSignupRegionsRequest(
  request: ListSignupRegionsRequest,
): string[] {
  const fields: string[] = [];
  if (!isCountryCode(request.resident_country)) fields.push("resident_country");
  if (
    request.pagination_key !== undefined &&
    !isPaginationKey(request.pagination_key)
  )
    fields.push("pagination_key");
  return fields;
}
export interface SignupRegion {
  tenant_id: string;
  hosting_country: CountryCode;
  hub_url: string;
  recommended: boolean;
}
export interface ListSignupRegionsResponse {
  catalog_version: string;
  regions: SignupRegion[];
  next_pagination_key: PaginationKey | null;
}
